test/libsolidity/smtCheckerTests/types/type_typehash.sol

```diff
@@ -0,0 +1,45 @@
+interface I1 {
+}
+
+contract C {
+    struct S1 {
+        uint256 x;
+    }
+
+    struct S2 {
+        uint256 x;
+        address y;
+    }
+
+    struct S3 {
+        uint256 x;
+        I1 y;
+        S2 third;
+    }
+
+    struct S4 {
+        S3 one;
+        S2 two;
+    }
+
+    struct S5 {
+        S2 two;
+        S1 one;
+        S3 three;
+        S4[5] four;
+    }
+
+	function f() public pure {
+		assert(type(S1).typehash == keccak256(bytes("S1(uint256 x)")));
+		assert(type(S2).typehash == keccak256(bytes("S2(uint256 x,address y)")));
+		assert(type(S3).typehash == keccak256(bytes("S3(uint256 x,address y,S2 third)S2(uint256 x,address y)")));
+        assert(type(S4).typehash == keccak256(bytes("S4(S3 one,S2 two)S2(uint256 x,address y)S3(uint256 x,address y,S2 third)")));
+        assert(type(S5).typehash == keccak256(bytes("S5(S2 two,S1 one,S3 three,S4[5] four)S1(uint256 x)S2(uint256 x,address y)S3(uint256 x,address y,S2 third)S4(S3 one,S2 two)")));
+	}
+}
+// ====
+// SMTEngine: all
+// ----
+// Warning 6328: (416-468): CHC: Assertion violation happens here.
+// Warning 6328: (503-555): CHC: Assertion violation happens here.
+// Info 1391: CHC: 5 verification condition(s) proved safe! Enable the model checker option "show proved safe" to see all of them.
```

Comment:
Why are they failing?

Reply:
I didn't found a way to check it locally, I use your CI :) Probably bug in implementation.

Reply:
For testing this. Once you build `isoltest`, run `./isoltest -t "*type_typehash*`. This is inside `build/test/tools`, where `build` is the cmake build directory.

Reply:
I've been debugging using `lldb`:

```bash
# from the root DIR
mkdir -p build
cmake -DCMAKE_BUILD_TYPE=Debug
make -C build -j2 solc isoltest

lldb -- ./build/solc/solc --ir "path.to.contract.sol"
# or
lldb -- ./build/test/tools/isoltest -t "path.to.test.file"
```

Reply:
```
build % ./test/tools/isoltest -t "*type_typehash*"
```
=>
```
Running tests...

smtCheckerTests/types/type_typehash.sol: NOT RUN

Summary: 0/7228 tests successful (7228 tests skipped).
```

Reply:
I think it's getting skipped because you compiled `isoltest` without Z3.

The following is probably what you used:
> cmake -DSTRICT_Z3_VERSION=OFF ..

Can you rebuild the build folder with `cmake ..` after installing the latest version of z3. Not sure about your platform, but what I like to do is checkout https://github.com/Z3Prover/z3 and the tag for latest release. Then the usual cmake build and `make install`.

Reply:
Also want to add, the encoding for SMT is the least important part here. We can even get this PR merged without supporting it. So feel free to ignore this until the others are done. I can also just fix it at the end myself.